### ISSUE TITLE
sync slow with large number of commits

### ISSUE THREAD

**Issue by ghost** [attic-labs/noms#2233]
The nomfs repo that @ahl linked to in #2220 has 20K+ commits. Syncing each of these commits is currently serialized because we need to each commit chunk to discover the ref of the previous one. We may need to add a server endpoint to do some of this work on the low-latency side of the connection

**Comment by aboodman** [comment]
Added format label because this probably implies changing the schema of `Commit` which may as well be a format change.

**Comment by aboodman** [comment]
Here is a simple and concrete proposal for how to fix this:

```
// There are two fairly distinct use cases for a Commit struct
//
// 1. Working with the current value
// 2. Exploring history
//
// This new Commit struct is optimized for use case (1). Both meta and history are
// out-of-line so that as much of the "meat" of the data for each commit is packed
// into the root chunk as possible.
struct Commit {
	meta?: Ref<Struct>
	history: Ref<Set<HistoryEntry>>
	value: Value
}

// To support use case (2), the concept of history is pulled out into its own
// separate struct with different inlining policy optimal for the use case of
// exploring history.
//
// Here, the data itself (commit) is out-of-line, but the parent history
// entries are inlined (and those entries' parents are also inlined, and
// so-on).
//
// The result is that as much of the history graph as possible is packed into
// one chunk. Each fetch gets you ~4k/20=204 nodes of the history graph instead
// of 1. The commits themselves still have to be fetched separately but that
// can be done in parallel.
struct HistoryEntry {
	// commit is out-of-line
	commit: Ref<Commit>
	// parent history entires are inline
	history: Set<HistoryEntry>
}
```

**Comment by rafael-atticlabs** [comment]
Is each `HistoryEntry.history` the set (i.e. usually just one) of immediate predecessors for the referenced commit? If not, how is the commit DAG navigated?

**Comment by aboodman** [comment]
Yes. Because everything is inlined and the probability of any individual
set item chunking still exists you should end up with tall skinny chunks of
about 4K I think.

**Comment by arv** [comment]
Seems promising to me

**Comment by rafael-atticlabs** [comment]
Right now a collection with 1 item will always inline item (it effectively can't chunk), so the common case of a lineage of commits each with one parent wouldn't ever chunk.

We could change chunking behavior so that a collection of one element would chunk.

**Comment by aboodman** [comment]
> We could change chunking behavior so that a collection of one element would chunk.

seems reasonable to me

**Comment by aboodman** [comment]
side note: I sorta feel like we should redefine "format change" to mean "if we make this change and you have an old noms codebase, it can't successfully talk to us anymore" - regardless of whether the format was actually changed.

**Comment by rafael-atticlabs** [comment]
Another possible solution:

```
struct Commit {
  meta?: Ref<Struct>
  parents: Set<Ref<Commit>>
  history: Ref<Map<Number, Set<Ref<Commit>>
}
```

The addition here is the reference to a new map which contains the full set of refs to commits *prior* to this one. The mapping is `height(Ref<Commit>) -> Set<Ref<Commits with that height>>`

This is useful for sync because the common thing to want is to quickly enumerate all commits between two points (a) -> (b). height(b) will always be > height(a), so just enumerating all commit refs with heights in that range will retrieve all of the commits you care about (it might return too many if there are branches which merge together, but it would frequently be exactly right.

Also, I think we could implement this *without* a noms version change. it's just an additional field on commit which sync, log or anything else can use if it's present to rapidly fetch a set of commits.

Lastly, for common cases of simple lineages of commits, the chunks in the map will dedup frequently because entries are effectively only ever added to the end of the map sequence.

**Comment by aboodman** [comment]
This is a pretty good idea for a quick workaround that doesn't require a format change. Cool. I guess whether to do this depends on how complicated it is to maintain the fork in the code that uses this if it is available or falls back to not using it.

**Comment by aboodman** [comment]
It's also worth remembering that your solution is asymptotically faster (and probably faster in practice) for the specific problem of sync, because it changes the problem from graph traversal to just an index scan.

**Comment by rafael-atticlabs** [comment]
New idea. Use a skip list. I.e. imagine using a prolly tree as an index. Observe that we only ever append to the end. If we use a prolly tree, we'll end up writing log n new chunks and reading log n chunks in order to mutate for each commit.

Using a skip list, means that we'll read (fewer than -on average- I think) log n chunks, but we write only the existing commit chunks.

We should be able to pick the level probability to have the same effective fan out as a prolly tree.
